compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.KtNodeTypes.REFERENCE_EXPRESSION
+import org.jetbrains.kotlin.config.LanguageFeature
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.declaration.isMemberExtension
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getLastChild
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.expressions.FirCallableReferenceAccess
+import org.jetbrains.kotlin.fir.expressions.FirGetClassCall
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.languageVersionSettings
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.resolve.containingClass
+import org.jetbrains.kotlin.fir.symbols.impl.FirBackingFieldSymbol
+
+object FirCallableReferenceChecker : FirQualifiedAccessChecker() {
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirCallableReferenceAccess) return
+        // [FirGetClassCallChecker] will check [FirGetClassCall].
+        if (expression is FirGetClassCall) return
+
+        checkReferenceIsToAllowedMember(expression, context, reporter)
+    }
+
+    // See FE 1.0 [DoubleColonExpressionResolver#checkReferenceIsToAllowedMember]
+    private fun checkReferenceIsToAllowedMember(
+        callableReferenceAccess: FirCallableReferenceAccess,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        // UNRESOLVED_REFERENCE will be reported separately.
+        if (callableReferenceAccess.calleeReference !is FirResolvedNamedReference) return
+        val reference = callableReferenceAccess.calleeReference as FirResolvedNamedReference
+
+        val source = reference.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+        // For example, rcv::member, reference source points to the entire callable reference.
+        // Note that both `rcv` and `member` would be of [REFERENCE_EXPRESSION] kind.
+        // We want to report on the last reference expression, i.e. the one after ::
+        val referenceSource = source.getLastChild(REFERENCE_EXPRESSION) ?: return
+
+        if (!context.session.languageVersionSettings.supportsFeature(LanguageFeature.CallableReferencesToClassMembersWithEmptyLHS)) {
```

Comment:
Please don't check for the features which appeared in 1.4 or earlier. FIR anyway shouldn't support so old stuff.

Reply:
(you may leave UNSUPPORTED_FEATURE itself in PR if you wish, it anyway will be usable for newer features)

Reply:
Okay, removed this check, but I'll leave `UNSUPPORTED_FEATURE`, which helped me figure out an issue in IDE generator.